### PR TITLE
Improved compiler typing

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Improved typing of compiler core

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] ~~Attached issue to pull request~~
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ]   ~~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )  ~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/execute/scheduler.py [hunk 1]
```diff
@@ -192,17 +208,20 @@ def define_types(self, compiler, statements, blocks):
...
 
         # first evaluate all definitions, this should be done in one iteration
         self.define_types(compiler, statements, blocks)
 
         # relations are also in blocks
-        statements = (s for s in statements if not isinstance(s, DefineRelation))
-        anchors = (
+        not_relation_statements: Generator[Statement, Any, Any] = (s for s in statements if not isinstance(s, DefineRelation))
+
+        anchors: Generator[Anchor, Any, Any] = (
```
Comment (sanderr) [inline src/inmanta/execute/scheduler.py:222]:
```suggestion
        not_relation_statements: Generator[Statement, None, None] = (s for s in statements if not isinstance(s, DefineRelation))

        anchors: Generator[Anchor, None, None] = (
```
From [the docs](https://docs.python.org/3/library/typing.html#typing.Generator): "If your generator will only yield values, set the `SendType` and `ReturnType` to `None` ... Alternatively, annotate your generator as having a return type of ... `Iterator[YieldType]`"

File: src/inmanta/compiler/__init__.py [hunk 2]
```diff
@@ -44,14 +44,19 @@
...
-from inmanta.plugins import PluginMeta
+from inmanta.plugins import Plugin, PluginMeta
 
 LOGGER = logging.getLogger(__name__)
 
+if TYPE_CHECKING:
+    from inmanta.ast import Statement, BasicBlock  # noqa: F401
+
 
 def do_compile(refs={}):
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/compiler/__init__.py:55]: done.

File: src/inmanta/execute/scheduler.py (hunk 1 above)
Comment (wouterdb) [reply src/inmanta/execute/scheduler.py:222]: thx, updated.

File: src/inmanta/compiler/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/compiler/__init__.py]: I forgot to do the top of the file, added that.

**Comment by wouterdb** [comment]
MyPy-diff

```diff
+ src/inmanta/ast/blocks.py:38: error: Incompatible types in assignment (expression has type "None", variable has type "ExecutionContext")
- src/inmanta/plugins.py:105: error: "Plugin" has no attribute "namespace"
- src/inmanta/plugins.py:201: error: "Plugin" has no attribute "namespace"
- src/inmanta/plugins.py:202: error: "Plugin" has no attribute "namespace"
- src/inmanta/plugins.py:203: error: "Plugin" has no attribute "namespace"
+ src/inmanta/execute/scheduler.py:77: error: "Type" has no attribute "get_all_instances"
+ src/inmanta/execute/scheduler.py:83: error: "Type" has no attribute "get_all_instances"
+ src/inmanta/execute/scheduler.py:89: error: "Type" has no attribute "get_all_instances"
+ src/inmanta/execute/scheduler.py:180: error: "DefinitionStatement" has no attribute "evaluate"
+ src/inmanta/execute/scheduler.py:221: error: "object" has no attribute "get_anchors"
+ src/inmanta/compiler/__init__.py:95: error: Incompatible types in assignment (expression has type "Optional[Namespace]", variable has type "Namespace")
+ src/inmanta/compiler/__init__.py:217: error: Incompatible types in assignment (expression has type "Optional[Namespace]", variable has type "Namespace")
+ src/inmanta/app.py:476: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[str, Type]")
+ src/inmanta/app.py:476: error: Incompatible types in assignment (expression has type "None", variable has type "Namespace")
+ src/inmanta/app.py:483: error: Argument 1 to "run" of "Exporter" has incompatible type "Dict[str, Type]"; expected "Optional[Dict[str, Entity]]"
+ src/inmanta/app.py:490: error: Argument 1 to "ModelExporter" has incompatible type "Dict[str, Type]"; expected "Dict[str, Entity]"
```

**Comment by sanderr** [comment]
GitHub makes it look really nice with all that green. Still, it's definitely progress.

File: src/inmanta/execute/scheduler.py [hunk 3]
```diff
@@ -192,17 +204,20 @@ def define_types(self, compiler, statements, blocks):
...
+    ) -> Sequence[Tuple[Location, Location]]:
         prev = time.time()
 
         # first evaluate all definitions, this should be done in one iteration
         self.define_types(compiler, statements, blocks)
 
         # relations are also in blocks
-        statements = (s for s in statements if not isinstance(s, DefineRelation))
-        anchors = (
+        not_relation_statements: Generator[Statement, Any, Any] = (s for s in statements if not isinstance(s, DefineRelation))
```
Comment (sanderr) [inline src/inmanta/execute/scheduler.py:216]:
```suggestion
        not_relation_statements: Iterator[Statement] = (s for s in statements if not isinstance(s, DefineRelation))
```
missed one

File: src/inmanta/execute/scheduler.py [hunk 4]
```diff
@@ -21,10 +21,10 @@
 import os
 import time
 from collections import deque
-from typing import Dict, List, Set, Tuple
+from typing import TYPE_CHECKING, Any, Deque, Dict, Generator, Iterator, List, Sequence, Set, Tuple
```
Comment (sanderr) [inline src/inmanta/execute/scheduler.py:24]:
```suggestion
from typing import TYPE_CHECKING, Any, Deque, Dict, Iterator, List, Sequence, Set, Tuple
```

File: src/inmanta/compiler/__init__.py [hunk 5]
```diff
@@ -116,53 +124,49 @@ class Compiler(object):
...
         """
-            This method will compile and prepare everything to start evaluation
+            This method will parse and prepare everything to start evaluation
             the configuration specification.
 
             This method will:
-            - load all namespaces
-            - compile the __config__ namespace
-            - start resolving it and importing unknown namespaces
+            - load all module using Project.get().get_complete_ast()
```
Comment (sanderr) [inline src/inmanta/compiler/__init__.py:167]:
```suggestion
            - load all modules using Project.get().get_complete_ast()
```

**Pull request merged**